core/vm/precompile_manager.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/precompile"
+)
+
+type methodID [4]byte
+
+type statefulMethod struct {
+	abiMethod     abi.Method
+	reflectMethod reflect.Method
+}
+
+type precompileMethods map[methodID]*statefulMethod
+
+type precompileManager struct {
+	evm         *EVM
+	precompiles map[common.Address]precompile.StatefulPrecompiledContract
+	pMethods    map[common.Address]precompileMethods
+}
+
+func NewPrecompileManager(evm *EVM) PrecompileManager {
+	precompiles := make(map[common.Address]precompile.StatefulPrecompiledContract)
+	pMethods := make(map[common.Address]precompileMethods)
+	return &precompileManager{
+		evm:         evm,
+		precompiles: precompiles,
+		pMethods:    pMethods,
+	}
+}
+
+func (pm *precompileManager) IsPrecompile(addr common.Address) bool {
+	_, isEvmPrecompile := pm.evm.precompile(addr)
+	if isEvmPrecompile {
+		return true
+	}
+
+	_, isStatefulPrecompile := pm.precompiles[addr]
+	return isStatefulPrecompile
+}
+
+func (pm *precompileManager) Run(
+	addr common.Address,
+	input []byte,
+	caller common.Address,
+	value *big.Int,
+	suppliedGas uint64,
+	readOnly bool,
+) (ret []byte, remainingGas uint64, err error) {
+
+	// run core evm precompile
+	p, isEvmPrecompile := pm.evm.precompile(addr)
+	if isEvmPrecompile {
+		return RunPrecompiledContract(p, input, suppliedGas)
+	}
+
+	contract, ok := pm.precompiles[addr]
+	if !ok {
+		return nil, 0, fmt.Errorf("no precompiled contract at address %v", addr.Hex())
+	}
+
+	// Extract the method ID from the input
+	methodId := methodID(input)
+	// Try to get the method from the precompiled contracts using the method ID
+	method, exists := pm.pMethods[addr][methodId]
+	if !exists {
+		return nil, 0, fmt.Errorf("no method with id %v in precompiled contract at address %v", methodId, addr.Hex())
+	}
+
+	gasCost := contract.RequiredGas(input)
+	if gasCost > suppliedGas {
+		return nil, 0, ErrOutOfGas
+	}
+
+	// Unpack the input arguments using the ABI method's inputs
+	unpackedArgs, err := method.abiMethod.Inputs.Unpack(input[4:])
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Convert the unpacked args to reflect values.
+	reflectedUnpackedArgs := make([]reflect.Value, 0, len(unpackedArgs))
+	for _, unpacked := range unpackedArgs {
+		reflectedUnpackedArgs = append(reflectedUnpackedArgs, reflect.ValueOf(unpacked))
+	}
+
+	ctx := precompile.NewStatefulContext(pm.evm.StateDB, addr, caller, value)
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && !ctx.IsReadOnly() {
+		ctx.SetReadOnly(true)
+		defer func() { ctx.SetReadOnly(false) }()
+	}
+
+	results := method.reflectMethod.Func.Call(append(
+		[]reflect.Value{
+			reflect.ValueOf(contract),
+			reflect.ValueOf(ctx),
+		},
+		reflectedUnpackedArgs...,
+	))
+
+	// check if precompile returned an error
+	if len(results) > 0 {
```

Comment:
can we check that the result parameters (length, type) are what are expected based on the method signature if possible?

Reply:
makes sense, will add